daemon/stop.go

```diff
@@ -122,5 +125,9 @@ func (daemon *Daemon) containerStop(ctx context.Context, ctr *container.Containe
 		// container did exit, so ignore previous errors and continue
 	}
 
+	// Ensure container status changes are committed by handler of container exit before returning control to the caller
+	ctr.Lock()
+	defer ctr.Unlock()
```

Comment:
Looks like this caused a regression; see https://github.com/docker/compose/pull/12950#issuecomment-2980759296

The `ctr.Lock()` makes this function wait for the container to die, but because it produces the `stop` event in a `defer`, the `stop` event now gets produced _AFTER_ the `die` event;

https://github.com/moby/moby/blob/8e6cd44ce42962a53e16f138c0d04bc8de346d1f/daemon/stop.go#L74-L78

Reply:
opened a PR;

- https://github.com/moby/moby/pull/50220